src/net/zmq.cpp

```diff
@@ -0,0 +1,184 @@
+                default:
+                    /* zmq is using cerrno errors. C++ spec indicates that
+                       `std::errc` values must be identical to the cerrno value.
+                       So just map every zmq specific error to the generic errc
+                       equivalent. zmq extensions must be in the switch or they
+                       map to a non-existent errc enum value. */
+                    return std::errc(value);
+                }
+                return std::error_condition{value, *this};
+            }
+
+        };
+        static const category instance{};
+        return instance;
+    }
+
+    void terminate::call(void* ptr) noexcept
+    {
+        assert(ptr != nullptr); // see header
+        while (zmq_term(ptr))
+        {
+            if (zmq_errno() != EINTR)
+                break;
+        }
+    }
+
+    namespace
+    {
+        //! RAII wrapper for `zmq_msg_t`.
+        class message
+        {
+            zmq_msg_t handle_;
+
+        public:
+            message() noexcept
+              : handle_()
+            {
+                zmq_msg_init(handle());
+            }
+
+            message(message&& rhs) = delete;
+            message(const message& rhs) = delete;
+            message& operator=(message&& rhs) = delete;
+            message& operator=(const message& rhs) = delete;
+
+            ~message() noexcept
+            {
+                zmq_msg_close(handle());
+            }
+
+            zmq_msg_t* handle() noexcept
+            {
+                return std::addressof(handle_);
+            }
+
+            const char* data() noexcept
+            {
+                return static_cast<const char*>(zmq_msg_data(handle()));
+            }
+
+            std::size_t size() noexcept
+            {
+                return zmq_msg_size(handle());
+            }
+        };
+
+        struct do_receive
+        {
+            /* ZMQ documentation states that message parts are atomic - either
+               all are received or none are. Looking through ZMQ code and
+               Github discussions indicates that after part 1 is returned,
+               `EAGAIN` cannot be returned to meet these guarantees. Unit tests
+               verify (for the `inproc://` case) that this is the behavior. 
+               Therefore, read errors after the first part are treated as a
+               failure for the entire message (probably `ETERM`). */
+            int operator()(std::string& payload, void* const socket, const int flags) const
+            {
+                int read = 0;
+                message part{};
+                for (;;)
+                {
+                    if ((read = zmq_msg_recv(part.handle(), socket, flags)) < 0)
+                        break;
+
+                    payload.append(part.data(), part.size());
+                    if (!zmq_msg_more(part.handle()))
+                        break;
+                }
+                return read;
```

Comment:
Should this not accumulate the >0 returns from all the zmq_msg_recv calls ? From the man page:
> The zmq_msg_recv() function shall return number of bytes in the message if successful.

Reply:
The caller is only checking for non-negative values, so theres no harm in dropping the information in its current use. But it might be bad conceptually. One issue is the possibility of overflow. Internally ZMQ clamps the value to `int::max`, meaning the return bytes are already dubious in edge cases. Which is why `zmq-msg_size(...)` has to be used. I'll duplicate that behavior here - accumulate but clamp to max value in the return.